src/Repository/RepoFileManager.cpp

```diff
@@ -0,0 +1,153 @@
+#include "Util/ConvertString.hpp"
+#include "Util/HexString.hpp"
+
+#define LOCAL_HASHES_FILE "local_hashes"
+
+void 
+RepoFileManager::Load(Path repofilepath)
+{
+  repository.Clear();
+
+  FileLineReaderA reader(repofilepath);
+  bool success = ParseFileRepository(repository, reader);
+
+  if (!success) {
+    repository.Clear();
+    //TODO: throw here?
+  }
+
+  LoadLocalHashes();
+}
+
+void
+RepoFileManager::Download(const char *name)
+{
+  const AvailableFile *remote_file = repository.FindByName(name);
+  if (remote_file == nullptr)
+    return;
+
+  const UTF8ToWideConverter base(remote_file->GetName());
+  if (!base.IsValid())
+    return;
+
+  Net::DownloadManager::Enqueue(remote_file->GetURI(), Path(base));
+
+  if (remote_file->HasHash()) {
+    local_hashes[remote_file->name] = remote_file->sha256_hash;
+  }
+}
+
+bool
+RepoFileManager::IsOutOfDate(const char *name)
+{
+  const AvailableFile *remote_file = repository.FindByName(name);
+
+  if (remote_file == nullptr)
+    return false;
+
+  if (remote_file->HasHash()) {
+    try {
+      std::array<std::byte, 32> local_hash = local_hashes[name];
+
+      return local_hash != remote_file->sha256_hash;
+    } catch (std::exception &e) {
+      // No hash of the local version was found so mark as out of date.
+      return true;
+    }
+  }
+
+#ifdef HAVE_POSIX
+  BrokenDate remote_changed = remote_file->update_date;
+
+  const UTF8ToWideConverter base(remote_file->GetName());
+  if (!base.IsValid())
+    return false;
+
+  BrokenDate local_changed = static_cast<BrokenDate>(BrokenDateTime::FromUnixTimeUTC(
+        File::GetLastModification(LocalPath(base))));
+
+  return local_changed < remote_changed;
+#else
+  return false;
+#endif
+}
+
+
+/**
+ * Reads the hashes of files from the repo from the disk.
+ * Every line is an entry of the form `filename = hash`.
+ * This is done so we don't have to calculate the
+ * checksum of the local files on every start.
+ */
+void
+RepoFileManager::LoadLocalHashes() {
+  FileLineReaderA linereader(LocalPath(LOCAL_HASHES_FILE));
+  KeyValueFileReader kvreader(linereader);
+
+  KeyValuePair kv_pair;
+  while (kvreader.Read(kv_pair)) {
+    std::string name = kv_pair.key;
+    try {
+      std::array<std::byte, 32> sha256 = ParseHexString<32>(kv_pair.value);
+
+      local_hashes[name] = sha256;
+    } catch (std::exception &e) {
+      // invalid hash in the hash file don't add anything
+    }
+  }
+}
+
+/**
+ * Writes the hashes of the files from the repo to the disk.
+ */
+void
+RepoFileManager::SaveLocalHashes() {
+  FileOutputStream file_os(LocalPath(LOCAL_HASHES_FILE),
+                           FileOutputStream::Mode::CREATE_VISIBLE);
+  BufferedOutputStream buf_os(file_os);
+  KeyValueFileWriter kvwriter(buf_os);
+
+  for (const auto &pair : local_hashes) {
+    std::string name(pair.first);
+    std::string hash(RenderHexString(pair.second));
+    kvwriter.Write(name.c_str(), hash.c_str());
+  }
+
+  buf_os.Flush();
```

Comment:
The `FileOutputStream` needs to be committed, or else the file will be deleted by its constructor.